### PR TITLE
feat(server): migrate RuntimeConfigStore to PostgreSQL (ADR-0060)

### PR DESCRIPTION
**Pull request by FortitudeEtc**
## Summary

Migrates `RuntimeConfigStore` (persistent runtime-config overrides: retention windows, `log_level`, DEX alert-routing knobs, OIDC settings) from SQLite to PostgreSQL, per ADR-0060 — closing the migration ladder to one remaining store (`OffloadTargetStore`/ADR-0059).

- **Per-key secrecy, not per-column**: `is_secret_config_key()` decides, at the row level, which of two tables a key lives in — `runtime_config` (plain) or `runtime_config_secrets` (`SecretCodec`-envelope-encrypted `sealed_value`). The OIDC client secret is no longer stored plaintext at rest.
- `read_secret()` returns the zeroizing `SecureBuffer` type (ADR-0010 §1), never `std::string` — a governance-blocking finding on the store's first draft.
- `get_all()`/`get()`/`read_secret()` all read both tables in ONE `UNION ALL` statement (one MVCC snapshot) — no window for a concurrent `set()` to make a key transiently vanish from any read.
- `set()`/`remove()` on a secret key take a transaction-scoped `pg_advisory_xact_lock` keyed by the secret key, closing a chaos-confirmed concurrent clear-vs-set race.
- No SQLite backfill (ADR-0009's fresh-start-by-default amendment) — a boot-time detect-and-warn check logs a row count if a legacy `runtime-config.db` still has real overrides.
- `/healthz` and `/readyz` both now report this store; `GET`/`PUT /api/config` 503 on a genuine database error instead of a response that reads as "nothing configured."

**Breaking**: existing `runtime-config.db` overrides (including the OIDC client secret) do NOT carry over on this cutover. Operators must reapply Settings overrides once after upgrading — see `docs/user-manual/upgrading.md`.

## Governance

Full `/governance` pipeline run across two rounds:

1. **Initial migration** (commits `c57932c24`..`766e1320b`): standard 8-gate pipeline, iterated to a clean bill. Findings recorded in `governance.d/2758-redact-runtime-config-secrets.6MqD21.jsonl` and earlier fragments.
2. **Post-rebase hardening round** (this PR's `a4c3c899f`/`e999ca6e5`): a fresh full run after rebasing onto `origin/dev` surfaced two BLOCKING findings — a `read_secret()` TOCTOU race (independently found by unhappy-path, chaos-injector, and enterprise-readiness) and an unjoined-thread hazard in two new regression tests (independently found by cpp-safety, cpp-expert, and quality-engineer). Both fixed and **empirically verified**: cpp-safety fault-injected the exact thread-hazard regression and confirmed a clean red assertion instead of `std::terminate`; quality-engineer mutation-tested the new crypto-error coverage (reverted the fix, watched it redden, restored it, watched it green). The `read_secret()` fix itself is structurally argued (one statement = one MVCC snapshot), not race-tested — same precedent as `get()`/`get_all()`'s prior identical fix. Gate 6's `sre` seat was satisfied by Gate 3's `sre` pass on this same store (identical scope, no separate re-run). Gate 8 re-review (7 agents: security-guardian, cpp-safety, cpp-expert, quality-engineer, sre, docs-writer, consistency-auditor) on the fix itself came back clean — no new BLOCKING/SHOULD. Full findings ledger: `governance.d/0060-runtime-config-store-hardening-round2.jkkfqg.jsonl` (22 rows: 17 fixed, 5 deliberately deferred pending follow-up issues — see below).

## Test plan

- [x] Full server + PostgreSQL test suite, all 15 shards, twice (before and after Gate 8 touch-ups): 0 failures, real multi-minute PG round-trips (not skip-mode).
- [x] `test_runtime_config_store.cpp` (new file): CRUD, degrade-typed reads, detect-and-warn, becomes-secret-later transitional state, advisory-lock serialization (genuine-block proof via a held second connection, not timing luck), decrypt-failure surfaced as `unexpected` (mutation-proved).
- [x] `test_runtime_config_secret_redaction.cpp` (new file): redaction contract.
- [x] `docs/test-coverage.md` updated with both new test files.

## Follow-ups

5 non-blocking items deliberately deferred (ledger rows `rtc2-12/13/14/15/22`, disposition `open`) — being filed as separate issues alongside this PR.

Co-Authored-By: Claude Sonnet 5 <noreply@anthropic.com>

### CODE REVIEW

**Comment by FortitudeEtc** [comment]
Pushed follow-up commit \`15b9633a0\`: superseded the 5 deferred (\`open\`) ledger rows now that the follow-up issues are filed — #3633, #3634, #3635, #3636, #3637. No code changes in this push, ledger-only.

File: server/core/src/runtime_config_store.cpp [hunk 1]
```diff
@@ -271,19 +664,122 @@ std::expected<void, std::string> RuntimeConfigStore::set(const std::string& key,
...
+                    legacy_db_path.string());
+        return;
+    }
+    const auto count = static_cast<std::size_t>(sqlite3_column_int64(stmt.get(), 0));
+    if (count > 0)
+        spdlog::warn("RuntimeConfigStore: legacy {} holds {} override(s) that will NOT be "
+                    "carried over (ADR-0009 fresh-start-by-default) -- reapply them via "
+                    "Settings after this boot",
+                    legacy_db_path.string(), count);
 }
```
Comment (Doomgoose) [inline server/core/src/runtime_config_store.cpp:837]: issue(blocking): this function opens the retained legacy `runtime-config.db` read-only and counts rows, but never hardens its permissions (or its `-wal`/`-shm` sidecars) to owner-only - unlike `webhook_store.cpp:958-1003`'s precedent for the identical scenario (an ADR-0010 obligation from a prior governance finding on PR #3563). The pre-migration store ran `PRAGMA journal_mode=WAL` unconditionally, so a real deployment that ever set `oidc_client_secret` likely has a `-wal` sidecar independently holding the plaintext value, in addition to the main file. Found independently by both adversarial reviewers (Kimi, Codex) and reproduced a third time by the orchestrator (`umask 022` + WAL mode -> 0644 on all three files). Fix: copy `webhook_store.cpp`'s `#ifndef _WIN32` permissions block (~25 lines) into this function, covering the main file and both sidecars.

File: server/core/src/runtime_config_store.cpp [hunk 2]
```diff
@@ -237,29 +565,94 @@ std::expected<void, std::string> RuntimeConfigStore::set(const std::string& key,
...
+                    std::vector<std::string>{key, sanitized_by, std::to_string(now)});
+                return up.status() == PGRES_COMMAND_OK;
+            });
+            if (!ok)
+                return std::unexpected(std::string(kRuntimeConfigDbErrorPrefix) +
+                                       "database write failed");
+        } else {
+            // ADR-0010 encrypt-failure semantics: encrypt OUTSIDE any lease,
+            // BEFORE any write. A failed encrypt touches nothing.
+            const auto* bytes = reinterpret_cast<const std::uint8_t*>(sanitized_value.data());
```
Comment (Doomgoose) [inline server/core/src/runtime_config_store.cpp]: issue(blocking): `sanitized_value` (computed unconditionally two lines above via `sanitize_pg_text(value)`) is what gets encrypted here, not the raw `value` - silently replacing any embedded NUL byte or invalid-UTF-8 sequence in the OIDC client secret before it's ever sealed. `sealed_value` is `BYTEA NOT NULL`, so this sanitization serves no storage purpose on the secret path. `webhook_store.cpp:544-547` and `plugin_config_store.cpp:326-329` (this PR's own cited ADR precedents) both encrypt the raw plaintext directly. A rotated credential containing such a byte reports `applied: true` but silently stores the wrong value - OIDC token exchange then fails against the IdP with no error surfaced anywhere in the chain. Fix: sanitize only in the non-secret branch (or only `sanitized_by`, which does go into TEXT columns); encrypt raw `value` bytes directly.

**Review by Doomgoose — CHANGES_REQUESTED** [review]
**Verdict:** REQUEST CHANGES - three confirmed blockers, all secret-handling; the migration's core structure and two governance-tracked race fixes are otherwise sound
**Tally:** 3 blocking - 5 important - 5 minor - 1 trivial - 1 suggestion

## Summary

- 11 lenses + a genuine two-reviewer adversarial pass (Kimi K2.7 + Codex, fully reachable, no quota issue this time) all converged: the two governance-tracked BLOCKING race fixes from this PR's own hardening rounds (`read_secret()` TOCTOU via UNION ALL, advisory-lock clear-vs-set) are genuinely fixed - independently re-verified by three separate reviewers reading the code, plus a live run of 41 test cases/412 assertions against real Postgres.
- Three NEW blockers surfaced, none previously caught by this PR's own two internal governance rounds: (1) the new `runtime_config_secret_codec_` is never enrolled in `kek_enrolled_codecs()` - found independently by two internal lenses (spec, history) and confirmed by me directly against source - so KEK rotation silently skips the OIDC secret while `/kek/status` falsely reports complete; (2) the retained legacy `runtime-config.db` and its WAL/SHM sidecars are never hardened to 0600, unlike this repo's own `WebhookStore` precedent for the identical scenario - found independently by both external adversarial reviewers and reproduced a third time by the orchestrator; (3) `set()` runs the OIDC client secret through `sanitize_pg_text()` before encrypting it, silently corrupting any embedded NUL byte or invalid-UTF-8 sequence in the credential - contradicting the sibling-store precedent (WebhookStore, PluginConfigStore) this PR's own ADR cites, which encrypt raw.
- All three are narrow, precedented fixes (the WAL/SHM fix is a ~25-line copy from `webhook_store.cpp`; the KEK fix is one line; the sanitize fix is reordering two existing lines) - none require redesign.
- Five further important, non-blocking findings filed as follow-up issues.

## What I verified (and did not)

- Read deeply: full scoped diff of `runtime_config_store.cpp/.hpp` (~1100 combined lines), `server.cpp`'s construction/teardown/KEK/health/REST-handler diffs, the 455-line ADR-0060, all 11 of this PR's own commits (`git show` each), and the governance ledger's 22+5 rows
- Ran: an adversarial pass (Kimi K2.7 + Codex, both full dynamic) compiled the PR head clean, stood up a real Postgres 18.6 cluster, and ran 41 test cases/412 assertions live (including both genuine `pg_locks`-polling concurrency tests) - all green; independently reproduced the WAL/SHM world-readable-file exposure via `umask 022` + `PRAGMA journal_mode=WAL`. I independently confirmed the KEK-enrollment gap and the sanitize-before-encrypt corruption directly against source myself before drafting this review.
- Not reviewed: a live fault-injection (revert-and-watch-it-redden) mutation test on either of the two already-fixed governance-tracked races - structural code reading (converging across three independent reviewers) plus the passing genuine-concurrency tests are the basis for calling both fixed

## Blocking - must fix before merge

1. issue(blocking) - `runtime_config_secret_codec_` (constructed at `server.cpp:5627`) is never added to `kek_enrolled_codecs()` (`server.cpp:21364-21377`), unlike the two sibling per-store codecs that are. KEK rotation/rewrap/status all iterate this one list - rotating the KEK silently never re-wraps the OIDC secret, while `/kek/status` reports `rotation_complete: true`. If the old key is later retired per standard rotation hygiene, the OIDC secret becomes permanently undecryptable, which per this PR's own ADR fails the entire server boot. Found independently by two lenses (spec, history), confirmed directly against source.
2. issue(blocking) - `warn_if_legacy_data_present()` opens the retained legacy `runtime-config.db` read-only and counts rows, but never hardens its permissions or checks for `-wal`/`-shm` sidecars - unlike `webhook_store.cpp:958-1003`'s exact precedent for this same scenario (an ADR-0010 obligation from a prior governance finding on PR #3563). The pre-migration store ran `PRAGMA journal_mode=WAL` unconditionally, so a real deployment that ever set `oidc_client_secret` likely has a `-wal` sidecar independently holding the plaintext value. Found independently by both external adversarial reviewers, reproduced a third time by the orchestrator (`umask 022` + WAL journal mode -> 0644 on all three files).
3. issue(blocking) - `set()` computes `sanitized_value = sanitize_pg_text(value)` unconditionally, then encrypts `sanitized_value` (not the raw `value`) on the secret-key path - silently replacing any embedded NUL byte or invalid-UTF-8 sequence in the OIDC client secret before it's ever sealed. `sealed_value` is `BYTEA`, so this sanitization serves no storage purpose on the secret path. `WebhookStore`/`PluginConfigStore` (this PR's own cited precedent) both encrypt raw. A rotated credential containing such a byte would report success while silently storing the wrong value, failing OIDC token exchange later with no error anywhere in the chain.

## Important - should fix

1. issue(important) - a filesystem error checking whether the legacy file exists is misclassified identically to "genuinely absent," with zero logging - could hide a real un-migrated plaintext secret under a permission-drift scenario with no warning at all. Tracked in #3639.
2. issue(important) - the breaking non-carryover of legacy config (including the OIDC secret) has no API-visible signal anywhere, only a boot-time log line the PR's own docs candidly admit is dashboard-invisible. Tracked in #3640.
3. issue(important) - `security-hardening.md`'s "current encryption posture" page - the exact page this PR's own ADR tells operators to consult - omits the legacy-plaintext-residue caveat disclosed everywhere else in this same PR's docs. Tracked in #3641.
4. issue(important) - `docs/adr/0010`'s secrets table still lists webhook signing secrets as plaintext, stale since ADR-0057 merged; this PR touched the adjacent row and fixed the identical stale claim in a sibling doc without carrying the fix here. Tracked in #3642.
5. issue(important) - `runtime_config_secrets` has no independent `has_secret` flag (row presence is the entire signal), so a SQL-write actor could force a downgrade or inject unauthenticated plaintext - found independently by both adversarial reviewers, banded below blocking by the orchestrator since it needs raw DB write access and shares its design with already-shipped `PluginConfigStore` (not unique to this PR). Tracked in #3643.

## Minor & Trivial

- minor - shutdown resets the shared `auth_key_provider_` roughly 80 lines before `runtime_config_store_`/`runtime_config_secret_codec_`, leaving the codec's borrowed `KekProvider&` dangling for that interval - structural debt, no demonstrated use-after-free, new to this PR.
- minor - no concurrency/fault-injection test proves the `read_secret()` UNION ALL fix specifically (unlike the two advisory-lock tests, which genuinely spawn a thread and poll `pg_locks`) - the fix is structurally sound (one statement, one MVCC snapshot) but rests on code review, not an executed race, for this one function.
- minor - `/healthz`/`/readyz` report this store healthy purely from construction success, not live read health - an existing, uniformly-applied convention shared by ~15 sibling stores, not introduced or worsened by this PR.
- minor - `PUT /api/config`'s DB/crypto failures reclassified from 400 to 503 with a new nested error envelope - a real wire-contract change for an existing error case, but deliberate and already documented in upgrading.md/rest-api.md.
- minor - `docs/capability-map.md` has the same plaintext-residue caveat gap as security-hardening.md, lower stakes as a terse capability index.
- trivial - GET's 503 message text changed wording ("runtime configuration store" -> "runtime config store"), intentionally unifying with PUT's wording; documented as a deliberate drift-fix.
- suggestion - the 3-line advisory-lock acquire+check block is duplicated verbatim at three call sites in this new code (no established multi-call-site precedent to follow, unlike several other apparent duplications in the file that turned out to be genuine repo-wide conventions) - worth one small helper.

## Follow-ups

- Legacy-file existence-check error swallowed - tracked in #3639
- Breaking change invisible to any API consumer - tracked in #3640
- security-hardening.md missing plaintext-residue caveat - tracked in #3641
- ADR-0010 stale webhook-plaintext claim - tracked in #3642
- No independent has_secret flag (shared design with PluginConfigStore) - tracked in #3643
- Already correctly tracked by this PR itself: #3633-#3637 (5 deliberately-deferred governance items)

## Process notes

- Re-review triggered on the next push - I'll verify all three blockers' fixes directly and re-approve once addressed.

**Comment by FortitudeEtc** [comment]
Pushed 3 new commits addressing all three BLOCKING findings from the review above, plus a scoped-governance follow-up round:

**`9ad2f9b4f`** — fixes the 3 blocking findings directly:
1. `runtime_config_secret_codec_` enrolled in `kek_enrolled_codecs()` (was missing — KEK rotation was silently never re-wrapping the OIDC secret while `/kek/status` reported `rotation_complete: true`).
2. `warn_if_legacy_data_present()` now hardens the legacy `runtime-config.db` and its `-wal`/`-shm` sidecars to 0600 before opening, porting `webhook_store.cpp`'s exact precedent for the identical scenario.
3. `set()` now encrypts the raw secret value instead of a `sanitize_pg_text()`'d copy — `sealed_value` is BYTEA, so sanitizing served no storage purpose and was silently corrupting an embedded NUL/invalid-UTF-8 byte before sealing.

New regression tests for #2 and #3 (mutation-proved). #1 has no new automated test — `kek_enrolled_codecs()` is a private `ServerImpl` member with no live-Session-free construction path, a pre-existing architecture-wide gap `test_operator_surface_twins.cpp` already documents, not something this fix introduces.

**`0610efa34`** — a scoped 7-agent governance pass on the fix above (security-guardian, docs-writer, cpp-safety, cpp-expert, architect, sre, quality-engineer; all PASS, no BLOCKING) surfaced and closed two more: a teardown-ordering hazard (the codec was reset in `stop()` *after* the KeyProvider it borrows), and dead code (`sanitized_value` computed unconditionally even where the raw-encrypt fix made it unused) plus an assertion-ordering nit in the new test.

**`8fdd696b8`** — docs: the legacy-file hardening fix made 5 doc claims stale ("never mutated"/"left in place, untouched" — now split into content-vs-permission-bits), plus a stale "THIRD SecretCodec instance" count corrected to fourth, plus 3 new checklist items added to `docs/postgres-store-playbook.md` (KEK-enrollment, legacy-file hardening, encrypt-raw) so this exact pattern doesn't recur on `OffloadTargetStore`/ADR-0059, the one store left on the ladder.

Verified: full server+PG test suite, 15/15 shards, 0 failures, after each commit.

Three non-blocking follow-ups from the governance pass filed as separate issues rather than folded into this PR: #3644 (orphaned-sidecar edge case, inherited from `WebhookStore`'s own precedent), #3645 (no boot-time observability if a future codec is omitted from `kek_enrolled_codecs()` again — this exact class of omission has now shipped twice), #3646 (`runtime_config_secret_codec_` not wired to `set_audit_hook`, unlike its `auth`/`webhook` siblings — may be an intentional exception, may be an oversight).

File: server/core/src/runtime_config_store.cpp (hunk 1 above)
Comment (FortitudeEtc) [reply server/core/src/runtime_config_store.cpp:837]: Fixed in `9ad2f9b4f`: ported `webhook_store.cpp`'s exact `#ifndef _WIN32` permissions block — forces 0600 on the main file and both `-wal`/`-shm` sidecars before `sqlite3_open_v2`. New regression test (`test_runtime_config_store.cpp`) mirrors `test_webhook_store.cpp`'s analogous WAL/SHM 0600 test — creates group/world-readable dummy sidecars, calls the function, asserts owner-only perms on all three files afterward.

File: server/core/src/runtime_config_store.cpp (hunk 2 above)
Comment (FortitudeEtc) [reply server/core/src/runtime_config_store.cpp]: Fixed in `9ad2f9b4f`: the secret-key path now encrypts raw `value` directly, not `sanitized_value`. New regression test round-trips a value with an embedded NUL + an invalid-UTF-8 byte through `set()`/`read_secret()` and asserts byte-for-byte equality — mutation-proved (reverting the fix reddens it with a visible U+FFFD/NUL-corrupted mismatch). Also folded in a follow-up from our own scoped-governance re-review (`0610efa34`): `sanitized_value`'s computation moved into the non-secret branch entirely, since the raw-encrypt fix left it as dead, unwiped work on the secret path.

**Comment by FortitudeEtc** [comment]
@Doomgoose — replied inline on the two line-anchored blockers (#2 legacy-file hardening, #3 raw-encrypt). For the third (KEK-enrollment, no line anchor since it's in `server.cpp`'s `kek_enrolled_codecs()` rather than in the diff you were reviewing): also fixed in `9ad2f9b4f` — `runtime_config_secret_codec_` is now enrolled, matching `webhook_secret_codec_`'s existing entry exactly, so it participates correctly in rotate/rewrap/status and the decrypt-failure metrics export loop.

All three confirmed fixed against source before touching anything, and covered by the full green test run noted in the earlier comment. Ready for your re-review whenever you get to it.

**Review by Doomgoose — APPROVED** [review]
**Verdict:** APPROVE - all three blockers confirmed fixed, plus a genuinely thorough follow-up round
**Tally:** 0 blocking - 5 important (unchanged, non-blocking) - 5 minor - 1 trivial - 1 suggestion

## Re-review of the three fix commits (`9ad2f9b4f`, `0610efa34`, `8fdd696b8`)

Verified each of the three prior blockers directly against the new diff:

1. C1 (KEK-enrollment gap) - fixed correctly. `runtime_config_secret_codec_` is now pushed into `kek_enrolled_codecs()`, matching `webhook_secret_codec_`'s exact pattern. No new automated regression exists for this specifically - honestly disclosed as a pre-existing, architecture-wide gap (`kek_enrolled_codecs()` has no live-Session-free construction path for any of its entries), not something introduced or left untested by this fix.
2. C2 (WAL/SHM sidecar hardening) - fixed correctly and thoroughly. `warn_if_legacy_data_present()` now chmods the main file AND both `-wal`/`-shm` sidecars to 0600, with honest error logging if a chmod fails - ported directly from `webhook_store.cpp`'s precedent. A new regression test verifies all three files land at the correct mode.
3. C3 (sanitize-before-encrypt corruption) - fixed correctly. `set()` now encrypts the raw `value` (not `sanitized_value`) on the secret-key path, and the empty-vs-set branch gate was switched to check the real input too. A new round-trip test proves a NUL/invalid-UTF-8 byte survives encrypt-then-decrypt intact.

## Beyond the three blockers - a genuinely thorough follow-up round

- A scoped 7-agent governance pass on the fix itself caught a real shutdown-ordering issue: `runtime_config_store_`/`runtime_config_secret_codec_` were being reset AFTER `auth_key_provider_`, even though the codec borrows a reference to it - the exact same defect class (KeyProvider-reference-outlives-teardown-order) I flagged on WebhookStore earlier in this review cycle. Moved both resets earlier, verified correctly ordered now, matching `webhook_secret_codec_`'s position. Currently unreachable (no live caller exists in that window) but correctly fixed as structural debt rather than left as a live risk.
- Dead-code cleanup: `sanitized_value` was still computed unconditionally even on the now-raw-encrypt secret path: moved into the only branch that still needs it.
- Docs: five stale claims the fix itself introduced ("never mutated"/"left in place, untouched" no longer fully true once permissions ARE now touched; a THIRD->FOURTH SecretCodec instance count correction) were caught and corrected across ADR-0060, the migration ladder, upgrading.md, and the SOC2 doc - genuine self-correction, not something I had to catch.
- `docs/postgres-store-playbook.md` gained 3 new checklist items (KEK-enrollment, legacy-file+sidecar hardening, encrypt-raw-never-sanitized) aimed squarely at `OffloadTargetStore`/ADR-0059, the one store left on the migration ladder - a real process improvement from this review cycle, not just a point fix.

## Still open (non-blocking, unchanged from prior review)

The five important-tier findings from the original review remain open and correctly untouched by this fix round (they're independent of the three blockers): #3639 (legacy-file existence-check error swallowed), #3640 (breaking change invisible to API), #3641 (security-hardening.md missing caveat), #3642 (ADR-0010 stale webhook claim), #3643 (no independent has_secret flag). None of these block merge.

## Process notes

- auto-merge armed - merges when required checks pass
- monitor active - pushes dismiss this approval and I re-approve automatically
